fluent/src/bundle.js

```diff
@@ -194,50 +172,141 @@ export default class FluentBundle {
   /**
    * Format a message to a string or null.
    *
-   * Format a raw `message` from the bundle into a string (or a null if it has
-   * a null value).  `args` will be used to resolve references to variables
-   * passed as arguments to the translation.
+   * Find a message or an attribute by `path` in the bundle and format it into
+   * a string (or a null if it is a message and has a null value). `path` may
+   * be a simple message identifier (`foo`) or a path to an attribute using
+   * a dot as the separator (`foo.bar`).
+   *
+   * If the `path` is a simple identifier and the message is not found in the
+   * bundle, this method returns `undefined`. If the `path` is an attribute
+   * path and the message doesn't have any attributes or the requested
+   * attribute is not found, it returns `undefined` too.
+   *
+   * `args` can be an object or `null`. If it's an object, it will be used to
+   * resolve references to variables passed as arguments to the translation.
    *
    * In case of errors `format` will try to salvage as much of the translation
    * as possible and will still return a string.  For performance reasons, the
    * encountered errors are not returned but instead are appended to the
    * `errors` array passed as the third argument.
    *
-   *     const errors = [];
-   *     bundle.addMessages('hello = Hello, { $name }!');
-   *     const hello = bundle.getMessage('hello');
-   *     bundle.format(hello, { name: 'Jane' }, errors);
-   *
-   *     // Returns 'Hello, Jane!' and `errors` is empty.
+   *     let errors = [];
+   *     bundle.addMessages('hello = Hello, {$name}!');
    *
-   *     bundle.format(hello, undefined, errors);
+   *     bundle.format('hello', {name: 'Jane'}, errors);
+   *     // → 'Hello, Jane!' and `errors` is empty.
    *
-   *     // Returns 'Hello, name!' and `errors` is now:
+   *     bundle.format('hello', undefined, errors);
+   *     // → 'Hello, name!' and `errors` is now:
+   *     // [<ReferenceError: Unknown variable: name>]
    *
-   *     [<ReferenceError: Unknown variable: name>]
+   *     errors.length = 0;
+   *     bundle.addMessages(`
+   *     email-input =
+   *         .placeholder = Your e-mail
+   *     `);
+   *     bundle.format('email-input.placeholder', null, errors);
+   *     // → 'Your e-mail' and `errors` is empty:
    *
-   * @param   {Object | string}    message
-   * @param   {Object | undefined} args
-   * @param   {Array}              errors
+   * @param   {string} path
+   * @param   {?Object} args
+   * @param   {?Array} errors
    * @returns {?string}
    */
-  format(message, args, errors) {
-    // optimize entities which are simple strings with no attributes
-    if (typeof message === "string") {
-      return this._transform(message);
+  format(path, args, errors) {
+    let parts = path.split(".");
```

Comment:
you could just use `path.split(".", 1);` to separate one `.` and not bother with anything past that. The rest is an attribute name.